tools/clang/tools/dxcompiler/dxcompilerobj.cpp

```diff
@@ -1564,13 +1564,17 @@ class DxcCompiler : public IDxcCompiler3,
 
     // only export shader functions for library
     compiler.getCodeGenOpts().ExportShadersOnly = Opts.ExportShadersOnly;
+    compiler.getLangOpts().ExportShadersOnly = Opts.ExportShadersOnly;
 
     if (Opts.DefaultLinkage.empty()) {
       compiler.getCodeGenOpts().DefaultLinkage = DXIL::DefaultLinkage::Default;
+      compiler.getLangOpts().DefaultLinkage = DXIL::DefaultLinkage::Default;
     } else if (Opts.DefaultLinkage.equals_lower("internal")) {
       compiler.getCodeGenOpts().DefaultLinkage = DXIL::DefaultLinkage::Internal;
+      compiler.getLangOpts().DefaultLinkage = DXIL::DefaultLinkage::Internal;
     } else if (Opts.DefaultLinkage.equals_lower("external")) {
       compiler.getCodeGenOpts().DefaultLinkage = DXIL::DefaultLinkage::External;
+      compiler.getLangOpts().DefaultLinkage = DXIL::DefaultLinkage::External;
```

Comment:
Alternatively, you could set one option from the other after this translation from string to enum. Like: `compiler.getLangOpts().DefaultLinkage = compiler.getCodeGenOpts().DefaultLinkage`.  Another option would be to change HLSLOptions to parse and store the enum value in the first place.